components/locid/src/zerovec.rs

```diff
@@ -19,24 +19,27 @@
 //! To perform lookup, store the stringified locale in a canonical BCP-47 form as a byte array,
 //! and then use [`Locale::strict_cmp()`] to perform an efficient, zero-allocation lookup.
 //!
+//! To produce more human-readable serialized output, you can use [`UnvalidatedStr`].
+//!
 //! ```
 //! use icu_locid::Locale;
 //! use std::str::FromStr;
 //! use zerovec::ZeroMap;
+//! use zerovec::ule::UnvalidatedStr;
 //!
 //! // ZeroMap from locales to integers
-//! let data: &[(&[u8], u32)] = &[
-//!     (b"de-DE-u-hc-h12", 5),
-//!     (b"en-US-u-ca-buddhist", 10),
-//!     (b"my-MM", 15),
-//!     (b"sr-Cyrl-ME", 20),
-//!     (b"zh-TW", 25),
+//! let data: &[(&UnvalidatedStr, u32)] = &[
+//!     (b"de-DE-u-hc-h12".into(), 5),
+//!     (b"en-US-u-ca-buddhist".into(), 10),
+//!     (b"my-MM".into(), 15),
+//!     (b"sr-Cyrl-ME".into(), 20),
+//!     (b"zh-TW".into(), 25),
 //! ];
-//! let zm: ZeroMap<[u8], u32> = data.iter().copied().collect();
+//! let zm: ZeroMap<UnvalidatedStr, u32> = data.iter().copied().collect();
 //!
 //! // Get the value associated with a locale
 //! let loc: Locale = "en-US-u-ca-buddhist".parse().unwrap();
-//! let value = zm.get_copied_by(|bytes| loc.strict_cmp(bytes).reverse());
+//! let value = zm.get_copied_by(|uvstr| loc.strict_cmp(&uvstr.0).reverse());
```

Comment:
`pub` tuple fields are not a great API, it'd be better to access this as either `uvstr.bytes`, or `uvstr.as_bytes()`.

Or, maybe that's nicest, implement `Deref<Target = [u8]>` on `UnvalidatedStr` so we can pass it in directly.

Reply:
I added `Deref` but `uvstr.cmp(some_bytes)` doesn't work. The following all work:

- `uvstr.0.cmp(some_bytes)`
- `some_bytes.cmp(uvstr).reverse()`
- `<[u8]>::cmp(uvstr, some_bytes)`
- `(uvstr as &[u8]).cmp(some_bytes)`

Preference?

Reply:
My preference would be `(**uvstr).cmp(some_bytes)`. I'd rather not keep the field public because it can be accessed with a deref (`*`) which doesn't require a name.

Reply:
I don't really agree with making the field private since there are no strong internal invariants, but one could argue that there's a soft invariant, so I did it. I also changed the last few places that were still using `.0` to use the deref operator instead.